core/state_transition.go

```diff
@@ -184,7 +184,7 @@ func (st *StateTransition) useGas(amount uint64) error {
 func (st *StateTransition) buyGas() error {
 	mgval := new(big.Int).Mul(new(big.Int).SetUint64(st.msg.Gas()), st.gasPrice)
 
-	if st.msg.GasCurrency() != nil && (st.gcWl == nil || !st.gcWl.IsWhitelisted(*st.msg.GasCurrency())) {
+	if st.msg.GasCurrency() != nil && (!currency.IsWhitelisted(*st.msg.GasCurrency(), nil, nil)) {
```

Comment:
1) Great catch - I'm fixing by passing in `st.state` to operate on pending state

2) This let me down a path that I'd love to get your input on. The fundamental question is whether we will have an issue if our EVM is built using a pending `stateDB` and the current header (the last finalized header). For this particular question I don't *think* this is an issue because I believe an `opSLoad` [operates on the stateDB](https://github.com/celo-org/celo-blockchain/blob/58a4453c22504e3367a4d95177320acda4c5b061/core/vm/instructions.go#L631) and disregards the header.

However, is there a situation where an inconsistency like this could cause a bug? The impact that the header has on the overall evm context comes from its use in setting
- BlockNumber
- Time
- Difficulty
- Hash
- GasLimit
which can be seen [here](https://github.com/celo-org/celo-blockchain/blob/58a4453c22504e3367a4d95177320acda4c5b061/core/evm.go#L67).

Not sure what risk this poses but could we see a state and header that are mismatched as faulty behavior somehow?

Reply:
1) If the change proposed [here](https://github.com/celo-org/celo-blockchain/pull/317#discussion_r307500130) is made, then I think passing in the header and state would be easy, since the `st` object has an vm.evm object attached to it.  I actually feel strongly about making sure there wouldn't be any inconsistencies in the block processing, since it's extremely important that all nodes process them in the exact same way.  We've been bitten before b/c of very subtle inconsistency differences, and those issues are either unreproducible and/or very hard to track down.

2) . Wouldn't we need evms built on in-process statedbs?  E.g. when a new block is currently being mined, the statedb will be in process, and not yet committed to the blockchain.

Reply:
This is done and I think you're right about in process states.